src/language_features/document_symbol.rs

```diff
@@ -23,6 +23,128 @@
+    ctx.call::<DocumentSymbolRequest, _>(
+        meta,
+        RequestParams::Each(req_params),
+        move |ctx: &mut Context, meta, results| {
+            match results.into_iter().find(|(_, v)| v.is_some()) {
+                Some((server_name, Some(DocumentSymbolResponse::Nested(symbols))))
+                    if !symbols.is_empty() =>
+                {
+                    let server = &ctx.language_servers[&server_name];
+                    let mut range_specs = String::default();
+                    indent_guides_calc_range_specs(
+                        &symbols,
+                        &params,
+                        &meta,
+                        ctx,
+                        server,
+                        &mut range_specs,
+                        0,
+                    );
+                    let version = meta.version;
+                    let command = formatdoc!(
+                        "set-option buffer lsp_indent_guides {version} {range_specs}
+                         set-option buffer lsp_indent_guides_timestamp {version}"
+                    );
+                    let command = format!(
+                        "evaluate-commands -buffer {} -- {}",
+                        editor_quote(&meta.buffile),
+                        editor_quote(&command)
+                    );
+                    ctx.exec(meta, command);
+                }
+                _ => {
+                    return;
+                }
+            };
+        },
+    );
+}
+
+fn indent_guides_calc_range_specs(
+    symbols: &[DocumentSymbol],
+    params: &IndentGuidesParams,
+    meta: &EditorMeta,
+    ctx: &Context,
+    server: &ServerSettings,
+    acc: &mut String,
+    level: usize,
+) {
+    for symbol in symbols {
+        let should_skip = level < params.skip;
+
+        let mut filename_path = PathBuf::default();
+        let filename = symbol_filename(meta, symbol, &mut filename_path).to_string();
+        let range = get_kakoune_range(server, filename.as_str(), &symbol.range, ctx).unwrap();
+
+        let children = symbol.children();
+
+        if !should_skip {
+            let is_not_in_child = children
+                .iter()
+                .find(|child| {
+                    let mut filename_path = PathBuf::default();
+                    let filename = symbol_filename(meta, *child, &mut filename_path).to_string();
+                    let range =
+                        get_kakoune_range(server, filename.as_str(), &child.range, ctx).unwrap();
+                    range.start.line <= params.position_line
+                        && params.position_line <= range.end.line
+                })
+                .is_none();
+
+            let is_inside =
+                range.start.line <= params.position_line && params.position_line <= range.end.line;
+            let is_focused = params.highlight_focused && is_not_in_child && is_inside;
+            let face = if is_focused {
+                "IndentGuideFocused"
```

Comment:
The `is_focused` logic doesn't seem to work for me.
I tried opening `src/text_edit.rs` and scroll around; indent guides work fine but they all have `IndentGuide` face.

In general, can we simplify this tree-walking so every node is only visited once?
We should be able to do this with a depth-first post-order traversal.
We always look at the innermost child first; and once we have found the symbol that "contains" the cursor, we ignore that check for all parents (or even for all other nodes).

```
def walk(node, position) -> bool:
    descendant_contains_position = False
    for child in node.children:
	descendant_contains_position |= walk(child, position)
    is_focused = not descendant_contains_position and node.contains(position)
    face = "IndentGuideFocused" if is_focused else "IndentGuide"
    return descendant_contains_position or is_focused
```

Reply:
Could it be something else? I'll try to simplify it anyway, yes, but does it work for you if try switching modes while being inside the innermost scope?

Could it be something with `NormalIdle` hook?

Reply:
Removed the timestamp logic, so it requests it all the time. Seems ok with the `--release` build. Wanted to have a smarter filtering logic like you've described and still can implement it, but decided to try the simpler solution first. Seems ok?